### PR TITLE
feat(sw): Add method to convert a htiled I1 buffer to vtiled

### LINKED ISSUE #7067: Color Format I1 vtile
### Problem to solve

Currently the I1 sw renderer assumes a htiled display. Either add a mode support of sorts to the renderer or provide method which takes a buffer and a temporary buffer and perform the conversion.

`lv_i1_convert_to_vtiled(uint8_t* buf, uint32_t width, uint32_t height, uint8_t* tmp, uint32_t tmp_buf_size)` 

Allowing the user to provide the temporary buffer allows for more flexibility on where to place the buffer. 

### Success criteria

Using vtiled monochrome displays is possible.

### Solution outline

See description 

### Rabbit holes

None that I know of?

### Testing

In CI using unit tests.

### Teaching

Using documentation in the spot where color formats are explained.

### Considerations

_No response_

### PR DESCRIPTION
**Pull request by faxe1008**
Adds a method to convert a monochrome draw buffer from htiled to vtiled orientation.

Resolves #7067.

A clear and concise description of what the bug or new feature is.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Review by liamHowatt — COMMENTED** [review]
Quick first impression: since it requires a temporary buffer, perhaps it would be better to simply make the temp buffer be the destination buffer and leave it up to the user to memcpy the dst back to the src on their own, if they choose to. I can imagine a user not wanting their source buffer to be overwritten in some cases.

**Comment by faxe1008** [comment]
Good call @liamHowatt will adjust it.

File: docs/porting/display.rst [hunk 1]
```diff
@@ -239,6 +239,36 @@ In full and direct modes, the buffer size should be large enough for the whole s
...
+
+.. code:: c
+
+    static void my_rounder_cb(lv_event_t *e)
+    {
+        lv_area_t *area = lv_event_get_param(e);
+
+        /* Round the width to the nearest multiple of 8 */
+        area->x1 = (area->x1 & ~0x7);
+        area->x2 = (area->x2 | 0x7);
```
Comment (kisvegabor) [inline docs/porting/display.rst]: This is done internally for I1 [here](https://github.com/lvgl/lvgl/blob/master/src/core/lv_refr.c#L294-L299).

File: docs/porting/display.rst [hunk 2]
```diff
@@ -239,6 +239,36 @@ In full and direct modes, the buffer size should be large enough for the whole s
 As LVGL can not handle fractional width make sure to round the horizontal resolution to 8-
 (For example 90 to 96)
 
+The :cpp:func:`lv_draw_sw_i1_convert_to_vtiled` function is used to convert a draw buffer in I1 color format from a row-wise (htiled) to a column-wise (vtiled) buffer layout.
+This conversion is necessary for certain display controllers that require a different draw buffer mapping. The function assumes that the buffer width and height are rounded to a multiple of 8.
+The bit order of the resulting vtiled buffer can be specified using the `bit_order_lsb` parameter.
+For more details, refer to the implementation in :cpp:func:`lv_draw_sw_i1_convert_to_vtiled` in :file:`src/draw/sw/lv_draw_sw.c`.
```
Comment (kisvegabor) [inline docs/porting/display.rst]: @vwheeler63 do we support the `:file:` directive?

Comment (vwheeler63) [reply docs/porting/display.rst]: > @vwheeler63 do we support the `:file:` directive?

Yes.  It is an "interpreted text role" built into Sphinx.  Here is its documentation:

image

Comment (kisvegabor) [reply docs/porting/display.rst]: Thank you!

**Comment by kisvegabor** [comment]
Please note the merge conflict.

**Comment by faxe1008** [comment]
Resolved the conflicts and addressed the documentation changes :^)

**Review by liamHowatt — DISMISSED** [review]
Looks solid. It took me a moment to understand the problem. The only other thing I'd like to see are more small tests just to cover width and height > 8 and `bit_order_lsb` `true`.

Nice pattern in the test buffer :^)
```
00111100
01000010
10000001
10100101
10000001
10000001
01000010
00111100
```

**Comment by kisvegabor** [comment]
@faxe1008 for the sake of completeness add a little bit more tests and we are good to go. :slightly_smiling_face:

File: tests/src/test_cases/draw/test_draw_sw_post_process.c [hunk 3]
```diff
@@ -315,4 +315,137 @@ void test_invert(void)
...
+    lv_draw_sw_i1_convert_to_vtiled(src_buf, 8, 8, 8, dst_buf, 8, false);
+    TEST_ASSERT_EQUAL_UINT8_ARRAY(expected_buf_msb, dst_buf, 8);
+
+    lv_draw_sw_i1_convert_to_vtiled(src_buf, 8, 8, 8, dst_buf, 8, true);
+    TEST_ASSERT_EQUAL_UINT8_ARRAY(expected_buf_lsb, dst_buf, 8);
+}
+
+void test_vtile_rectangular(void)
+{
+    uint8_t src_buf[64] = {
```
Comment (kisvegabor) [inline tests/src/test_cases/draw/test_draw_sw_post_process.c]:
```suggestion
    uint8_t src_buf[128] = {
```

**Review by kisvegabor — APPROVED** [review]
Looks good, thank you!

**Review by liamHowatt — APPROVED** [review]

**Pull request merged**